### ISSUE TITLE
Query against IOfflineTable<T> involving an enum property returns zero rows

### ISSUE THREAD

**Issue by oliverw** [Azure/azure-mobile-apps#610]
## Describe the bug

Given the following model:

```c#
namespace DataAccess.Model;

public enum TaskStatus
{
    Unknown = 0,
    Created = 1,
    Completed = 2,
}

public class Task : DatasyncClientData
{
    [JsonProperty("title")]
    public string Title  {get; set;}
    [JsonProperty("description")]
    public string Description  {get; set;}
    [JsonProperty("status")]
    public TaskStatus Status {get; set;}
    [JsonProperty("foo")]
    public int Foo {get; set;}
}
```

the following initialization ...

```c#
var options = new DatasyncClientOptions
{
    HttpPipeline = new []{ authService.RefreshTokenHandler },
    IdGenerator = (table) => AppConst.GenerateEntityId(),
    OfflineStore = SetupOfflineStore(authService.Ticket.UserId),
    ParallelOperations = 1,
};

var client = new DatasyncClient(appConfig.ApiBaseUrl, options);

await client.InitializeOfflineStoreAsync(ct);

Tasks = client.GetOfflineTable<Model.Task>();
```

and the following database contents ...

![image](https://user-images.githubusercontent.com/51688/218329825-00cab5af-70c3-4dac-b129-0760e695040f.png)

I get the following query results:

```c#
// IOfflineTable: 0 items returned, fail
var foo = await Tasks
    .Where(x => x.Status == DataAccess.Model.TaskStatus.Completed)
    .ToListAsync();
```
```c#
// IEnumerable: 1 items returned, pass
var bar = (await Tasks.ToListAsync())
    .Where(x => x.Status == DataAccess.Model.TaskStatus.Completed)
    .ToList();
```
```c#
// IRemoteTable: 1 items returned, pass
var baz = await Tasks.CreateQuery().RemoteTable
    .Where(x => x.Status == DataAccess.Model.TaskStatus.Completed)
    .ToListAsync();
```

## What platforms?

* Server (ASP.NET Framework, ASP.NET Core, NodeJS):
  * .NET 6
  * Postgres
  * Standalone

* Client (.NET, JavaScript, iOS, Android):
  * .NET 6 WinUI 3
  * Microsoft.Datasync.Client 5.0.21
  * Microsoft.Datasync.Client.SQLiteStore 5.0.21
  * Win11

**Comment by adrianhall** [comment]
The OfflineSQLiteStore is translating the WHERE clause to the following:

```text
SELECT * FROM [kitchensink] WHERE ([enumValue] = @p1)
@p1="Completed"
```

However, enums are stored as ints in the database.  In order to be backwards compatible, we need to convert to an int.  However, really, they should be stored as strings for forwards compatibility.  We'll fix the problem in 6.0.0 fully, but in the interim, we'll do the int conversion.

**Comment by oliverw** [comment]
> The OfflineSQLiteStore is translating the WHERE clause to the following:
> 
> ```
> SELECT * FROM [kitchensink] WHERE ([enumValue] = @p1)
> @p1="Completed"
> ```
> 
> However, enums are stored as ints in the database. In order to be backwards compatible, we need to convert to an int. However, really, they should be stored as strings for forwards compatibility. We'll fix the problem in 6.0.0 fully, but in the interim, we'll do the int conversion.

So whats your recommendation for now?

**Comment by adrianhall** [comment]
Actually - down the rabbit hole I go :-)

The root cause of this is that the ASP.NET Core serializer that we are using doesn't turn enums into strings - it turns them into ints instead (default enum handling).  I need to alter the service side serializer to emit enums as strings.

Not sure how I managed to miss that earlier, but I'm going to be adding a test for it as well now.

**Comment by adrianhall** [comment]
* Added test in Integration Tests / Server to explicit test round-tripping of enums as strings.
* Added a kitchensink test for enumValue searching in offline tables.
* Updated the JSON serializer in the service to use StringEnumConverter (which was the problem here)

Currently running tests, but it looks good, so I will have a PR in the next 30 minutes or so.  Once 5.1.1 comes out, you will need to resync any table that has an enum in it (otherwise the data will not be correct).

**Comment by oliverw** [comment]
> * Added test in Integration Tests / Server to explicit test round-tripping of enums as strings.
> * Added a kitchensink test for enumValue searching in offline tables.
> * Updated the JSON serializer in the service to use StringEnumConverter (which was the problem here)
> 
> Currently running tests, but it looks good, so I will have a PR in the next 30 minutes or so. Once 5.1.1 comes out, you will need to resync any table that has an enum in it (otherwise the data will not be correct).

Cool. Looking forward to it.

**Comment by oliverw** [comment]
Working fine.

**Comment by adrianhall** [comment]
That would not surprise me.  Don't do that :-)  We'll likely be able to fix that when we do 6.0.0

**Comment by oliverw** [comment]
> That would not surprise me. Don't do that :-) We'll likely be able to fix that when we do 6.0.0

Noted :)
